### PR TITLE
Chubin.ci

### PR DESCRIPTION
**Pull request by chubin**

### CODE REVIEW

**Comment by chubin** [comment]
@abitrolly It seems to work pretty good with Alpine here (the failing tests aside; they are broken everywhere). Are you sure we should switch to Ubuntu as suggested in #256?

**Comment by abitrolly** [comment]
@chubin I must admit that switching base for Docker image didn't fix anything. I will rework that PR.

**Comment by chubin** [comment]
@abitrolly I will try to fix these two failing tests in the meantime

File: Dockerfile [hunk 1]
```diff
@@ -4,8 +4,10 @@ FROM alpine:3.12
 RUN apk add --update --no-cache git py3-six py3-pygments py3-yaml py3-gevent \
       libstdc++ py3-colorama py3-requests py3-icu py3-redis
 ## building missing python packages
-RUN apk add --no-cache --virtual build-deps py3-pip g++ python3-dev \
-    && pip3 install --no-cache-dir rapidfuzz colored polyglot pycld2 \
+COPY requirements.txt /tmp
```
Comment (abitrolly) [inline Dockerfile:7]: Just moving these lines below to the top should do the trick.
```
WORKDIR /app
COPY . /app
```

Comment (chubin) [reply Dockerfile:7]: And after that `pip3 install -r /app/requirements.txt`, you mean?
Yes, sounds reasonable

Comment (abitrolly) [reply Dockerfile:7]: Yep.

**Comment by chubin** [comment]
@abitrolly We did it!

**Comment by abitrolly** [comment]
Interesting. The solution turns off caching by default. Maybe the caching system could include options as cache key? Then these queries will have separate cache entries.
```
python/copy+file?QT
python/copy+file?Q
python/copy+file
```
Here.
https://github.com/chubin/cheat.sh/blob/a4be5d8025686af8934ac22bb20f1e65a77a247c/lib/routing.py#L184-L186

**Comment by chubin** [comment]
Yes, it turns caching off, but only for one adapter: UpstreamAdapter. Basically, even in this case caching is needed, and it is a temporary workaround now, and not a real solution.

The problem here is that the whole post processing (adding comments, ansi rendering etc) are done by the upstream at the moment, which is not optimal and should be done on the downside instead, and the upstream should return raw data instead. I hope to fix it in future, and then this workaround can be removed.

Regarding your suggestion: maybe it is a good idea, and should be done indeed, but currently only expensive operations results are cached. Caching everything would simplify the whole architecture a lot. On the other hand, it would make possible situations when cache entries are inconsistent, but maybe it is not so bad.

Anyway, the whole caching architecture should be reworked in future, but there are some more acute problems, that must be fixed first.

For example, it would be good to change Dockerfile in such way so that the upstream repository data is created as a separate layer, and so it could be handled (updated, pulled, etc) semi-independently from the rest of the code

**Comment by abitrolly** [comment]
Thanks for the explanation. Figuring out what is downstream and what is upstream, as for me needs a diagram. Likely the subject of #231.

> For example, it would be good to change Dockerfile in such way so that the upstream repository data is created as a separate layer, and so it could be handled (updated, pulled, etc) semi-independently from the rest of the code

Oh yea. `lib` imports pull in too much stuff for external dependencies for the simple download script.

**Comment by chubin** [comment]
Yes, absolutely agree with it. I hope to do it after some code refactoring

**Review by sledzikteo — APPROVED** [review]

**Pull request merged**